docs/hot_module_reloading.md

```diff
@@ -0,0 +1,190 @@
+## What is Hot Module Reloading?
+
+This is a feature that allows you to speed up development process by reducing time between writing
+code and seeing the result. It detects changes in your code and automatically reloads minimum amount
+js modules on the page for changes to take effect.
+
+If something depends on changed module A, it may have some instances of A's types, functions and
+objects closured, so reloading A itself doesnt invalidate closure. In basic configuration
+hot-reloading reloads all modules in chain from changed module up to the root module of applications
+and reruns `main()` function for changes to take effect. This means that without special measures
+state of your won't be preserved, nevertheless hot-reloading is much faster than full page
+reloading.
+     
+## Turning hot-reloading on
+`build_runner` server has built-in support of hot reloading. To activate it just run `serve` command
+with `--hot-reload` option.
+
+## Using hooks to handle module reloading
+
+Each module where changes are detected is marked as invalidated. For each invalidated module three
+hooks will be called to determinate, is it possible to handle it's reloading, or all parents (other
```

Comment:
nit: determine if it is possible to handle the reload, or if all parents should be marked as invalidated too